cli/src/spinner.rs

```diff
@@ -0,0 +1,136 @@
+    where
+        F: Future,
+    {
+        Self::wrap_inner(future, None).await
+    }
+
+    /// Like `wrap` but also displays a message.
+    pub async fn wrap_with_message<F>(future: F, message: impl Into<String>) -> F::Output
+    where
+        F: Future,
+    {
+        Self::wrap_inner(future, Some(message.into())).await
+    }
+
+    async fn wrap_inner<F>(future: F, message: Option<String>) -> F::Output
+    where
+        F: Future,
+    {
+        let spinner = Spinner::new_inner(message);
+        let result = future.await;
+        spinner.stop().await;
+        result
+    }
+
+    /// Set a new message for the spinner to display.
+    pub async fn set_message(&self, message: impl Into<String>) {
+        self.tx
+            .send(Command::Message(Some(message.into())))
+            .await
+            .ok();
+    }
+
+    /// Remove the existing spinner message (if any)
+    pub async fn unset_message(&self) {
+        self.tx.send(Command::Message(None)).await.ok();
+    }
+
+    /// Stop the spinner. This requires a bit of cleanup, and so should be `await`ed before doing
+    /// any other i/o.
+    pub async fn stop(self) {
+        self.tx.send(Command::Stop).await.ok();
+        self.handle.await.ok(); // of no consequence if the spinner thread panics, squash it
+    }
+
+    /// Spin until receiver sees `Command::Stop` or the channel is closed.
+    async fn spin(mut rx: Receiver<Command>, msg: Option<String>) {
+        let mut msg = msg;
+        let mut dots = SPINNER_DOTS.iter().cycle();
+        eprint!("{}", ansi::CURSOR_POSITION_SAVE);
+        eprint!("{}", ansi::CURSOR_HIDE);
+        eprint!("{}", ansi::CLEAR_LINE);
+        let mut interval = tokio::time::interval(time::Duration::from_millis(SPINNER_DELAY));
+        loop {
+            match rx.try_recv() {
+                Err(TryRecvError::Empty) => {
+                    eprint!("{}", ansi::CURSOR_POSITION_0);
+                    eprint!("{}", ansi::CLEAR_LINE);
+                    eprint!("{}", dots.next().unwrap());
+                    if let Some(msg) = msg.as_ref() {
+                        eprint!(" {}", msg)
+                    }
+                    interval.tick().await;
+                }
+                Ok(Command::Message(new_msg)) => {
+                    msg = new_msg;
+                }
+                _ => {
+                    break;
+                }
+            }
+        }
+        eprint!("{}", ansi::CLEAR_LINE);
+        eprint!("{}", ansi::CURSOR_POSITION_RESTORE);
+        eprint!("{}", ansi::CURSOR_SHOW);
+    }
+}
+
+impl Default for Spinner {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+mod ansi {
+    pub const CURSOR_POSITION_SAVE: &str = "\x1B7";
+    pub const CURSOR_POSITION_RESTORE: &str = "\x1B8";
+    pub const CURSOR_POSITION_0: &str = "\x1B[0G";
+    pub const CURSOR_HIDE: &str = "\x1B[?25l";
+    pub const CURSOR_SHOW: &str = "\x1B[?25h";
+    pub const CLEAR_LINE: &str = "\x1B[2K";
+}
```

Comment:
I'm not sure how this worked in the spinners library, but having all these hardcoded **will** break on certain environments. Proper resolution of the `CURSOR_HIDE` escape sequence would involve resolving the `civis` terminfo capability for example.

Reply:
Spinners lib does the [same kinda thing](https://github.com/FGRibreau/spinners/blob/16ed2c6f5ec95c5cb9edca3e0f6debf144c7d78c/src/lib.rs#L192).

In another project I have used [crossterm](https://github.com/crossterm-rs/crossterm/blob/21155716e2eedd5ba8ab97168e5eed7cd50d2ad8/src/cursor.rs#L311-L320) which aims to have a much more varied support across terminals / OSes, but I was not sure if bringing in this dependency would be worthwhile.

Another note is that, while these now go to stderr, which is an improvement, another optimization would be to use [atty](https://docs.rs/atty/latest/atty/) and gate all prettifying ANSI codes based on tty status. For example `phs project > out` with no `-j` will currently print ANSI codes to the output file. Similarly `phs project > out 2>&1` will print the these spinner codes to the out file.

Reply:
I'd say checking for a TTY definitely makes sense, in fact this will likely be added to STD in a future release (https://github.com/rust-lang/rust/pull/98033).

With escape sequences there's unfortunately probably more people that do it wrong than people that do it right. I'm not sure it matters much for us since we have no goal to support a wide range of terminals, but it's certainly something to consider.

Reply:
So are there any changes you'd like me to make as part of this PR?

Reply:
I think a TTY check should be pretty trivial and provide the biggest benefit. Escapes I'd be okay with for now and put in a mental reminder should anyone ever complain that it looks like crap on the linux console or something.